### PR TITLE
CHECK that PreCheck and GetCheckUnit are not called twice

### PR DESCRIPTION
**Pull request by danakj**
If they were called twice for a CompilationUnit, they would destroy objects that they created and returned a pointer to, leaving a dangling pointer somewhere else.

### CODE REVIEW

File: toolchain/driver/compile_subcommand.cpp [hunk 1]
```diff
@@ -420,6 +420,8 @@ class CompilationUnit {
 
   auto PreCheck() -> Parse::NodeLocConverter& {
     CARBON_CHECK(parse_tree_, "Must call RunParse first");
+    CARBON_CHECK(!node_converter_.has_value(), "Called PreCheck twice");
```
Comment (jonmeow) [inline toolchain/driver/compile_subcommand.cpp]: In other calls (`RunLex`, `RunParse`, `PostCheck`, etc) we only verify that we're moving forward. They would share this problem. I'm fine with adding more checks to help document the progressive call structure, but I think it's implying the wrong thing to only check in some calls: can you please add them in the other spots too?

Comment (danakj) [reply toolchain/driver/compile_subcommand.cpp]: Right, these functions do not `emplace` any optionals, so it's not clear how to CHECK that without adding a bool for the purpose of checking, or maybe an enum tracking the progress through them all if it's linear. WDYT?

Comment (jonmeow) [reply toolchain/driver/compile_subcommand.cpp]: Note, we typically don't write `.has_value()` for optionals. `parse_tree_` is an `optional` [for example]. You could use the same CHECKs we're using to ensure things are called in order, to make sure they're not called again.

Comment (danakj) [reply toolchain/driver/compile_subcommand.cpp]: Oh, thanks, will do. The lack of has_value() and use of operator= had me miss that. There's a couple functions that are still missed, but this covers most of them (and all the ones that set optionals).

**Review by jonmeow — APPROVED** [review]
LG

**Pull request merged**